src/System.Private.CoreLib/src/System/Threading/Thread.CoreRT.cs

```diff
@@ -327,14 +327,121 @@ public bool Join(int millisecondsTimeout)
+            const int NsPerSecond = 1000 * 1000 * 1000;
+
+            ulong ticksPerSecond = HighPerformanceCounter.Frequency;
+            if (ticksPerSecond < 1000 / MeasureDurationMs)
+            {
+                // High precision clock not available or clock resolution is too low, resort to defaults
+                s_yieldsPerNormalizedYield = DefaultYieldsPerNormalizedYield;
+                s_optimalMaxNormalizedYieldsPerSpinIteration = DefaultOptimalMaxNormalizedYieldsPerSpinIteration;
+                return;
+            }
+
+            // Measure the nanosecond delay per yield
+            ulong measureDurationTicks = ticksPerSecond / (1000 / MeasureDurationMs);
+            uint yieldCount = 0;
+            ulong startTicks = HighPerformanceCounter.TickCount;
+            ulong elapsedTicks;
+            do
+            {
+                RuntimeImports.RhSpinWait(10);
+                yieldCount += 10;
+
+                ulong nowTicks = HighPerformanceCounter.TickCount;
+                elapsedTicks = nowTicks - startTicks;
+            } while (elapsedTicks < measureDurationTicks);
+            double nsPerYield = (double)elapsedTicks * NsPerSecond / ((double)yieldCount * ticksPerSecond);
+            if (nsPerYield < 1)
+            {
+                nsPerYield = 1;
+            }
+
+            // Calculate the number of yields required to span the duration of a normalized yield
+            int yieldsPerNormalizedYield = (int)(MinNsPerNormalizedYield / nsPerYield + 0.5);
+            if (yieldsPerNormalizedYield < 1)
+            {
+                yieldsPerNormalizedYield = 1;
+            }
+            else if (yieldsPerNormalizedYield > MaxYieldsPerNormalizedYield)
+            {
+                yieldsPerNormalizedYield = MaxYieldsPerNormalizedYield;
+            }
+
+            // Calculate the maximum number of yields that would be optimal for a late spin iteration. Typically, we would not want to
+            // spend excessive amounts of time (thousands of cycles) doing only YieldProcessor, as SwitchToThread/Sleep would do a
+            // better job of allowing other work to run.
+            int optimalMaxNormalizedYieldsPerSpinIteration =
+                (int)(NsPerOptimialMaxSpinIterationDuration / (yieldsPerNormalizedYield * nsPerYield) + 0.5);
+            if (optimalMaxNormalizedYieldsPerSpinIteration < 1)
+            {
+                optimalMaxNormalizedYieldsPerSpinIteration = 1;
+            }
+
+            s_yieldsPerNormalizedYield = yieldsPerNormalizedYield;
+            s_optimalMaxNormalizedYieldsPerSpinIteration = optimalMaxNormalizedYieldsPerSpinIteration;
+        }
+
+        public static void SpinWait(int iterations)
+        {
+            if (iterations <= 0)
+            {
+                return;
+            }
+
+            EnsureYieldProcessorNormalizedInitialized();
```

Comment:
Could consider doing the initialization at the start of the finalizer thread. It's a bit unfortunate as it increases CPU time during startup (though not wall clock time). The GC would need a place for it as well. In CoreCLR, before initialization the relevant values are default-initialized so that spin-waits can still work. Once the initialization happens it could inform the GC of the scaling factor.